# Data collection and description

### Leerdoelen 2. data exploratie & visualisatie
- Je leert welke onderdelen in een goede Data Understanding thuis horen.
- Je gaat zelf een rapport maken waar een goede Data Understanding in zit. 
- Je kan met Pandas data inlezen en bewerken. 
- Je kan met Seaborn of matplotlib visualisaties maken van de data. 

### Achtergrondinformatie werkcollege 1:
- Google naar 
    - `pandas read csv file`
- Datacamp course `Manipulating data with Pandas`

### Achtergrondinformatie werkcollege 1:
- Google naar 
    - `pandas dataframe make histogram`
    - `pandas correlation`
    - `seaborn correlation plot`
    - `seaborn pair plot`
    - `pandas dataframe plot variables`
- Datacamp course `Introduction to Data Visualisation with Python` en eventueel `Data Visualization with Seaborn`. 

### Opdracht 
- Maak een Data Understanding hoofdstuk voor de gegeven KNMI dataset, gemodelleerd naar het voorbeeld wat je gekregen hebt. De data is een bewerkte versie van data gedownload van [deze link](https://projects.knmi.nl/klimatologie/daggegevens/selectie.cgi).
- Leg uit (Trace&Explain) wat de code in `read_knmi_data` doet in elk van de stappen. 
- Als je een beeld van de data hebt, probeer iets te zeggen over het soort Business understanding wat je zou kunnen krijgen met deze data. 
- Data prepation, Modeling en Evaluation hoef je deze week niet aan te vullen. 

# Business understanding
Voor klimaatwetenschappers is het van groot belang om de ontwikkeling van het klimaat te volgen. Echte metingen kunnen trends laten zien, bijvoorbeeld een toename in het aantal extreme regenbuien. Een toename hiervan zou bijvoorbeeld gevolgen kunnen hebben voor de hoogte van de grote rivieren, en dus voor de benodigde hoogte van de dijken. 

De volgende dataset laat wat klimaatgegevens zien over de tijd. 

# Data Understanding
De data staat in een tekst bestand (.txt), maar na het openen van het bestand in een tekst editor is het duidelijk dat dit een csv-bestand is. De functie `read_knmi_data` leest de knmi datafile in.

## Instellingen aanpassen
Als je gewend bent aan andere ontwikkelomgevingen dan heb je bijv. gemist dat de beschikbare objecten getoond worden en dat je deze automatisch kunt aanvullen. Hiervoor stellen we de completer in op 'greedy' m.b.v. zogenaamde Jupyter Notebook Magic. 

Om later ook de visualisaties in het Notebook te tonen zorgen we dat de output van het package matplotlib ook als (inline) output wordt getoond in het Notebook.

In [10]:
#handige jupyter notebook magic
%config IPCompleter.greedy = True  
%matplotlib inline

## initial data collection

De data in het tekst-bestand is een csv-bestand. We kunnen dit direct inlezen als csv-bestand in het Notebook.

Hiervoor maken we gebruik van de read_csv methode in pandas. 
- de file moet je altijd meegeven
- de separator is ook handig om expliciet aan te geven. Doe je dit niet, dan wordt een komma gebruikt.
- In dit tekstbestand staat de toelichting op de variabelen als commentaar. Deze regels moeten worden genegeerd bij het inlezen van de gegevens. Dit doen we door dit mee te geven als argument in de methode.
- De kolomnamen staan ook als commentaar aangegeven. We hebben hier twee mogelijke werkwijzen: 1) het bronbestand aanpassen door voor deze regel het commmentaar aan te passen. Nadeel: het is dan later niet navolgbaar wat er is aangepast in de aangeleverde data. We kiezen voor 2) de gegevens inlezen zonder de header van het csv-bestand. Daarna geven we de kolommen in ons DataFrame begrijpbare namen. 

**We moeten één van onderstaande manieren kiezen om de gegevens te importeren. Welke vinden jullie het geschiktst? Licht je antwoord toe.**

In [2]:
#pandas importeren in onze werkomgeving
import pandas as pd
import os
import matplotlib as plt

In [12]:
#data files
path = os.path.normpath(os.getcwd())
filename= "data\knmi_data.txt"
data_source = os.path.join(path,filename)

knmi_data_1=pd.read_csv(data_source)
knmi_data_1

Hier kan errors voorkomen, alleen te gebruiken bij kleinere datasets waarvan jij 100% weet dat het structuur goed is.

In [13]:
knmi_data_2=pd.read_csv(data_source,comment='#', header=None)
knmi_data_2

0         1    2    3    4    5
0     278  20110201  -18  -38   11   15
1     278  20110202   29   11   40    3
2     278  20110203   52   30   80   14
3     278  20110204   90   61  109   16
4     278  20110205  114  107  120    0
...   ...       ...  ...  ...  ...  ...
3651  278  20210130   -4  -42   16    0
3652  278  20210131  -27  -64    7    0
3653  278  20210201    0   -9    9    2
3654  278  20210202   22    0   77   23
3655  278  20210203   81   61   95  183

[3656 rows x 6 columns]

Data in de datum kolom zijn niet geformateerd.

In [14]:
# parse_dates: Let pandas try and transform this column to a date
knmi_data_3=pd.read_csv(data_source,sep=',', comment='#', header=None, parse_dates=[1])
knmi_data_3

0          1    2    3    4    5
0     278 2011-02-01  -18  -38   11   15
1     278 2011-02-02   29   11   40    3
2     278 2011-02-03   52   30   80   14
3     278 2011-02-04   90   61  109   16
4     278 2011-02-05  114  107  120    0
...   ...        ...  ...  ...  ...  ...
3651  278 2021-01-30   -4  -42   16    0
3652  278 2021-01-31  -27  -64    7    0
3653  278 2021-02-01    0   -9    9    2
3654  278 2021-02-02   22    0   77   23
3655  278 2021-02-03   81   61   95  183

[3656 rows x 6 columns]

hiermee zijn data geformateerd, commentaar zijn afgehandeld en seperators zijn gespecificeerd

In [52]:
knmi_data_4=pd.read_csv(data_source,sep=',', comment='#', header=None, names=['station', 'datum', 'T_gem', 'T_min', 'T_max', 'Neerslag'], parse_dates=[1])

Timedelta('3655 days 00:00:00')

In [17]:
knmi_data_4=knmi_data_4.set_index(['station','datum'])

[plaats hier je feedback]

In [20]:
def read_knmi_data(filename, names):
    return pd.read_csv(filename, 
                       comment='#',               # Skip all comments
                       header=None,               # No header
                       names=names,
                       skipinitialspace=True,     # Fix the trailing spaces after the ','-separator
                       index_col=[0,1],           # Use the first two columns as the row index
                       parse_dates=[1])           # Let pandas try and transform the second column to a date

knmi_data_5 = read_knmi_data(data_source, names=['station', 'datum', 'T_gem', 'T_min', 'T_max', 'Neerslag'])
knmi_data_5

T_gem  T_min  T_max  Neerslag
station datum                                    
278     2011-02-01    -18    -38     11        15
        2011-02-02     29     11     40         3
        2011-02-03     52     30     80        14
        2011-02-04     90     61    109        16
        2011-02-05    114    107    120         0
...                   ...    ...    ...       ...
        2021-01-30     -4    -42     16         0
        2021-01-31    -27    -64      7         0
        2021-02-01      0     -9      9         2
        2021-02-02     22      0     77        23
        2021-02-03     81     61     95       183

[3656 rows x 4 columns]

[plaats hier je feedback]

[Onderbouw het alternatief dat jullie kiezen. Kopieer deze code en importeer de gegevens in onze werkomgeving als DataFrame knmi_data]

In [19]:
def read_knmi_data(filename, names):
    return pd.read_csv(filename,
                       comment='#',               # Skip all comments
                       header=None,               # No header
                       names=names,
                       skipinitialspace=True,     # Fix the trailing spaces after the ','-separator
                       index_col=[0,1],           # Use the first two columns as the row index
                       parse_dates=[1])           # Let pandas try and transform the second column to a date

knmi_data = read_knmi_data(data_source, names=['station', 'datum', 'T_gem', 'T_min', 'T_max', 'Neerslag'])

## data description
**Voer alle stappen uit die nodig zijn om:** 

*Describe the data that has been acquired, including the format of the data, the quantity of data (for example, the number of records and fields in each table), the identities of the fields, and any other surface features which have been discovered. Evaluate whether the data acquired satisfies the relevant requirements.* 
(bron: CRISP-DM1.0, p.18)

**Gebruik de 3 volgende cellen als format.**

[leg uit wat je gaat bekijken en waarom]

In [32]:
#code inclusief commentaar
knmi_data.shape


(3656, 4)

Ik zie de hoeveelheid kolommen en rijen, dat geeft mij inzicht in de grote van de data set en of er mogelijk wat ontbreekt

In [33]:
knmi_data.describe()

T_gem        T_min        T_max     Neerslag
count  3656.000000  3656.000000  3656.000000  3656.000000
mean    105.027899    59.338074   148.262856    20.448578
std      61.939013    56.406605    73.191528    41.534217
min    -121.000000  -179.000000   -54.000000    -1.000000
25%      60.000000    19.000000    92.000000     0.000000
50%     103.000000    60.500000   147.000000     1.000000
75%     153.000000   103.000000   204.000000    22.000000
max     280.000000   198.000000   393.000000   556.000000

Hier krijg ik statistische inzichten te zien voor ieder kolom die geen index is

In [34]:
knmi_data.head(5)

T_gem  T_min  T_max  Neerslag
station datum                                    
278     2011-02-01    -18    -38     11        15
        2011-02-02     29     11     40         3
        2011-02-03     52     30     80        14
        2011-02-04     90     61    109        16
        2011-02-05    114    107    120         0

Hiermee krijg ik inzicht in hoe de data eruit ziet, wat de waardes zijn en krijg ik een kleine overzicht of er mogelijke problemen of relaties zijn in mijn data.

In [53]:
max(knmi_data_4["datum"])- min(knmi_data_4["datum"])

Timedelta('3655 days 00:00:00')

Here we can see that the timedelta(measurement of time) of the dataset spans 3655 days.

In je toelichting beschrijf je in elk geval de variabelen in de dataset.

De variabelen beschrijven de volgende zaken:

- T_gem, gemiddelde temperatuur per dag
- T_min, de minimum temperatuur per dag
- T_max, de maximum temperatuur per dag
- Neerslag, de hoeveelheid neerslag per dag
- station, het stations id (in de index)
- datum, de datum (in de index)

[Aanvullen}